internal/querynodev2/delegator/idf_oracle.go

```diff
@@ -29,120 +35,279 @@ import (
-	RemoveGrowing(segmentID int64)
+	Register(segmentID int64, stats bm25Stats, state commonpb.SegmentState)
 
 	BuildIDF(fieldID int64, tfs *schemapb.SparseFloatArray) ([][]byte, float64, error)
-}
 
-type bm25Stats struct {
-	stats         map[int64]*storage.BM25Stats
-	activate      bool
-	targetVersion int64
-	snapVersion   int64
+	Start()
+	Close()
 }
 
-func (s *bm25Stats) Merge(stats map[int64]*storage.BM25Stats) {
+type bm25Stats map[int64]*storage.BM25Stats
+
+func (s bm25Stats) Merge(stats bm25Stats) {
 	for fieldID, newstats := range stats {
-		if stats, ok := s.stats[fieldID]; ok {
+		if stats, ok := s[fieldID]; ok {
 			stats.Merge(newstats)
 		} else {
-			s.stats[fieldID] = storage.NewBM25Stats()
-			s.stats[fieldID].Merge(newstats)
+			s[fieldID] = storage.NewBM25Stats()
+			s[fieldID].Merge(newstats)
 		}
 	}
 }
 
-func (s *bm25Stats) Minus(stats map[int64]*storage.BM25Stats) {
+func (s bm25Stats) Minus(stats bm25Stats) {
 	for fieldID, newstats := range stats {
-		if stats, ok := s.stats[fieldID]; ok {
+		if stats, ok := s[fieldID]; ok {
 			stats.Minus(newstats)
 		} else {
-			log.Panic("minus failed, BM25 stats not exist", zap.Int64("fieldID", fieldID))
+			s[fieldID] = storage.NewBM25Stats()
+			s[fieldID].Minus(newstats)
 		}
 	}
 }
 
-func (s *bm25Stats) GetStats(fieldID int64) (*storage.BM25Stats, error) {
-	stats, ok := s.stats[fieldID]
+func (s bm25Stats) GetStats(fieldID int64) (*storage.BM25Stats, error) {
+	stats, ok := s[fieldID]
 	if !ok {
 		return nil, errors.New("field not found in idf oracle BM25 stats")
 	}
 	return stats, nil
 }
 
-func (s *bm25Stats) NumRow() int64 {
-	for _, stats := range s.stats {
+func (s bm25Stats) NumRow() int64 {
+	for _, stats := range s {
 		return stats.NumRow()
 	}
 	return 0
 }
 
-func newBm25Stats(functions []*schemapb.FunctionSchema) *bm25Stats {
-	stats := &bm25Stats{
-		stats: make(map[int64]*storage.BM25Stats),
+type sealedBm25Stats struct {
+	sync.RWMutex // Protect all data in struct except activate
+	bm25Stats
+
+	activate *atomic.Bool
+
+	inmemory  bool
+	removed   bool
+	segmentID int64
+	ts        time.Time // Time of segemnt register, all segment resgister after target generate will don't remove
+	localPath string
+}
+
+func (s *sealedBm25Stats) writeFile(localPath string) (error, bool) {
+	s.RLock()
+
+	if s.removed {
+		return nil, true
 	}
 
+	m := make(map[int64][]byte, len(s.bm25Stats))
+	stats := s.bm25Stats
+	s.RUnlock()
+
+	// RUnlock when stats serialize and write to file
+	// to avoid block remove stats too long when sync distribution
+	for fieldID, stats := range stats {
+		bytes, err := stats.Serialize()
+		if err != nil {
+			return err, false
+		}
+
+		m[fieldID] = bytes
+	}
+
+	b, err := json.Marshal(m)
+	if err != nil {
+		return err, false
+	}
+
+	err = os.WriteFile(localPath, b, 0600)
+	if err != nil {
+		return err, false
+	}
+	return nil, false
+}
+
+func (s *sealedBm25Stats) ShouldToLocal() bool {
```

Comment:
nit: rename this function to `ShouldOffLoadToDisk`, and add comment explaining:

After merged the stats of a segment into the overall stats, Delegator still need to store the segment stats, so that later when the segment is removed from target, we can Minus its stats. To reduce memory usage, idfOracle store such per segment stats to disk, and load them when removing the segment.